### DISCUSSION TITLE
Where does command argument path expansion happen?

### DISCUSSION THREAD

**Discussion by CD3** [xonsh/xonsh#5113] [Q&A] [answered]
I would like to write an extension to enhance the command line argument expansion, specifically for path arguments. I have been looking through the Xonsh source code and can not figure out where this get done.

For example, when user types

```
cd ~/Downloads
```

The `~/Downloads` argument get turned into `/home/user/Downloads`. But where does this happen?

cdclark

**Comment by anki-code** [comment, marked as answer]
Parser:

https://github.com/xonsh/xonsh/blob/ed213f852b963a2dfad96414c462d688e93c792a/xonsh/parsers/base.py#L3478-L3484

Injection:

https://github.com/xonsh/xonsh/blob/359b4b2ab1836549d5296d217e341214a6ab1751/xonsh/built_ins.py#L536

Implementation:

https://github.com/xonsh/xonsh/blob/359b4b2ab1836549d5296d217e341214a6ab1751/xonsh/tools.py#L105-L122

**Comment by CD3** [comment]
Perfect, thank you.
